src/orion/algo/space/configspace.py

```diff
@@ -0,0 +1,261 @@
+
+        for _, dim in space.items():
+            cdim = self.visit(dim)
+
+            if cdim:
+                dims.append(cdim)
+
+        cspace.add_hyperparameters(dims)
+        return cspace
+
+
+def to_configspace(space: Space) -> ConfigurationSpace:
+    """Convert orion space to configspace
+
+    Notes
+    -----
+    ``ConfigurationSpace`` will set its own default values
+    if not set inside ``Space``
+
+    """
+    conversion = ToConfigSpace()
+    return conversion.space(space)
+
+
+@singledispatch
+def to_oriondim(dim: Hyperparameter) -> Dimension:
+    """Convert a config space hyperparameter to an orion dimension"""
+    raise NotImplementedError()
+
+
+@to_oriondim.register
+def from_categorical(dim: CategoricalHyperparameter) -> Dimension:
+    """Builds a categorical dimension from a categorical hyperparameter"""
+    choices = {k: w for k, w in zip(dim.choices, dim.probabilities)}
+    return Categorical(dim.name, choices)
+
+
+@to_oriondim.register(UniformIntegerHyperparameter)
+@to_oriondim.register(UniformFloatHyperparameter)
+def from_uniform(dim: Hyperparameter) -> Dimension:
+    """Builds a uniform dimension from a uniform hyperparameter"""
+
+    klass = Integer
+    args = []
+    kwargs = dict(
+        # NOTE: Config space always has a config value
+        # so orion-space would get it as well
+        default_value=dim.default_value
+    )
+
+    if isinstance(dim, UniformFloatHyperparameter):
+        klass = Real
+    else:
+        kwargs["precision"] = int(-log10(dim.q)) if dim.q else 4
+
+    dist = "uniform"
+    args.append(dim.lower)
+    args.append(dim.upper)
+
+    if dim.log:
+        dist = "reciprocal"
+
+    return klass(dim.name, dist, *args, **kwargs)
+
+
+@to_oriondim.register(NormalFloatHyperparameter)
+@to_oriondim.register(NormalIntegerHyperparameter)
+def from_normal(dim: Hyperparameter) -> Dimension:
+    """Builds a normal dimension from a normal hyperparameter"""
+
+    klass = Integer
+    args = []
+    kwargs = dict(
+        # NOTE: Config space always has a config value
+        # so orion-space would get it as well
+        default_value=dim.default_value
+    )
+
+    if isinstance(dim, NormalFloatHyperparameter):
+        klass = Real
+    else:
+        kwargs["precision"] = int(-log10(dim.q)) if dim.q else 4
+
+    dist = "norm"
+    args.append(dim.mu)
+    args.append(dim.sigma)
+
+    if dim.lower:
+        kwargs["low"] = dim.lower
+        kwargs["high"] = dim.upper
+
+    return klass(dim.name, dist, *args, **kwargs)
+
+
+@to_orionspace.register
+def configspace_to_orionspace(cspace: ConfigurationSpace) -> Space:
+    """Convert from orion space to configspace
+
+    Notes
+    -----
+    ``ConfigurationSpace`` will set default values for each dimensions of ``Space``
+
+    """
+    space = Space()
+
+    for _, cdim in cspace.get_hyperparameters_dict().items():
+        odim = to_oriondim(cdim)
+        print(cdim, odim)
+        assert odim is not None
```

Comment:
Should raise a ValueError() with an explicit error message.

Reply:
Should it ? If the dimension is not supported `NotImplementedError` will be raised.
`to_oriondim()` cannot return None

Reply:
Yes I'm not asking that. I'm asking, why is the `odim` asserted to not be None, if you're saying that it can never be None?
Could it be because you use something like `Optional[Dimension]` as the return annotation of `to_oriondim`? If so, then the annotation should be `Dimension`, and the assert isn't necessary.

Reply:
It documents the invariant that odim cannot be none, i.e "stating the fact" as well as checking it is

Reply:
I'm a huge proponent of asserts btw, unlike @bouthilx  that has a bit of Assertophobia :P

There is no need to state this fact here. This doesn't add any information that wasn't already encoded by the type hints. You can remove this assert.

Reply:
Asserts don't check things. They affirm things. IF `odim` could possibly be `None`, that's different. You'd want to do as @bouthilx  says and raise a ValueError. However, your type-signature for `to_oriondim` already indicates it can't be None. The assert here is not needed.

Reply:
fixed

Reply:
@lebrice he-hum, Assertinmaincodeophobia. :stuck_out_tongue: